records/2024-02-12.grouping-sms-codes.md

```diff
@@ -0,0 +1,123 @@
+restricted to two-factor authentication only, the software needs a mechanism
+to associate it only with templates of 2FA nature, i.e. the short code meant
+to send 2FA should be used to send 2FA SMS.
+
+At the current time, because the short code is sitting along long codes and all
+of these are in a default pool, that means that all codes are randomly used to
+send SMS, including the short codes. Hence the latter can be used to send
+SMS messages that are non-related with 2FA.
+
+To overcome this issue:
+
+- we need to isolate the short code and long codes within their own group that will be identified with proper usage,
+- we need a way to identify the templates with two-factor authentication.
+
+The ability to create a group of codes has extra benefits that we can leverage
+on top of solving the current issue:
+
+- the ability to have a formal random pool of long codes,
+- the ability to assign multiple dedicated long codes to one service,
+- the ability to repatriate the existing dedicated long codes from the us-west-1 region into the ca-central-1 and align better with our own policies of sending from Canada exclusively.
+
+### Overhaul of SMS sending
+
+This fits into our vision for a general overhaul of how we send SMS. The
+following areas along with broad tasks are envisioned at the time of this
+writing:
+
+```mermaid
+mindmap
+  root((short code))
+    Manage keywords
+    ::icon(fa-solid fa-key)
+        Identify replied service
+        Manage internal service block lists
+            Store block list per service
+            Report on the block list per service
+            Remove number from block list
+    Associate 2FA with templates
+    ::icon(fa-solid fa-link)
+        Provide usage on template form
+        ::icon(fa-brands fa-wpforms)
+        Identify existing 2FA template
+        ::icon(fa-solid fa-magnifying-glass)
+    Sending 2FA
+    ::icon(fa fa-comment-sms)
+        Update logic for original numbers
+        ::icon(fa-solid fa-code)
+            Fetch code with template
+            Fetch code with service
+            Fallback to default pool
+        Send SMS through Pinpoint
+        Create pool for short code
+    Repatriate dedicated codes to Canada
+    ::icon(fa-brands fa-canadian-maple-leaf)
+        Create pool for existing services with dedicated code
+        Auto pool creation on code assignment
+        Auto pool destruction on code removal
+```
+
+## Options
+
+Below are the different options that are considered to group SMS codes
+together.
+
+### Option 1: Leverage AWS Pinpoint pools
+
+Create a phone pool in AWS Pinpoint for regular long codes which act as a
+fallback. Create a pool containing the short code. Resurrect the Pinpoint
+implementation to send SMS and leverage the botocore Pinpoint API to send
+messages identified with a specific phone pool.
+
+#### Pros
+
+- Fast to implement
+
+#### Cons
+
+- Still cannot track which numbers that were used.
```

Comment:
Pinpoint returns the `originationPhoneNumber` in the delivery receipt 🎉

Reply:
Nice! I will remove the con here and potentially add a note. Do you know if the origination phone number is also delivered in the current receipts when using SNS? I wonder if that is simply a lack of coverage from our part now.

Reply:
No, currently we just see the destination in there unfortunately.
```
{
    "notification": {
        "messageId": "9c186762-79c7-5741-8361-39f10506b1ba",
        "timestamp": "2024-04-12 18:16:33.482"
    },
    "delivery": {
        "phoneCarrier": "Bell Cellular Inc. / Aliant Telecom",
        "mnc": 610,
        "numberOfMessageParts": 1,
        "destination": "+16135550123",
        "priceInUSD": 0.01181,
        "smsType": "Transactional",
        "mcc": 302,
        "providerResponse": "Message has been accepted by phone",
        "dwellTimeMs": 0,
        "dwellTimeMsUntilDeviceAck": 521
    },
    "status": "SUCCESS"
}

Reply:
Ah we got the price in there too! good to know! 👍